contracts/HATVault.sol

```diff
@@ -0,0 +1,770 @@
+    function getArbitrator() public view returns(address) {
+        if (arbitrator != NULL_ADDRESS) {
+            return arbitrator;
+        } else {
+            return registry.defaultArbitrator();
+        }
+    }
+
+    /** @notice See {IHATVault-getChallengePeriod}. */
+    function getChallengePeriod() public view returns(uint256) {
+        if (challengePeriod != NULL_UINT) {
+            return challengePeriod;
+        } else {
+            return registry.defaultChallengePeriod();
+        }
+    }
+
+    /** @notice See {IHATVault-getChallengeTimeOutPeriod}. */
+    function getChallengeTimeOutPeriod() public view returns(uint256) {
+        if (challengeTimeOutPeriod != NULL_UINT) {
+            return challengeTimeOutPeriod;
+        } else {
+            return registry.defaultChallengeTimeOutPeriod();
+        }
+    }
+
+    /* -------------------------------------------------------------------------------- */
+
+    /* --------------------------------- Helpers -------------------------------------- */
+
+    /**
+    * @dev Deposit funds to the vault. Can only be called if the committee had
+    * checked in and deposits are not paused.
+    * @param caller Caller of the action (msg.sender)
+    * @param receiver Reciever of the shares from the deposit
+    * @param assets Amount of vault's native token to deposit
+    * @param shares Respective amount of shares to be received
+    */
+    function _deposit(
+        address caller,
+        address receiver,
+        uint256 assets,
+        uint256 shares
+    ) internal override virtual nonReentrant {
+        if (!committeeCheckedIn)
+            revert CommitteeNotCheckedInYet();
+        if (shares == 0) revert AmountToDepositIsZero();
+        if (withdrawEnableStartTime[receiver] != 0 && receiver == caller) {
+            // clear withdraw request
+            withdrawEnableStartTime[receiver] = 0;
+        }
+
+        super._deposit(caller, receiver, assets, shares);
+    }
+
+    function _withdraw(
+        address caller,
+        address receiver,
+        address owner,
+        uint256 assets,
+        uint256 shares,
+        uint256 fee
+    ) internal nonReentrant {
+        if (assets == 0) revert WithdrawMustBeGreaterThanZero();
+        if (caller != owner) {
+            _spendAllowance(owner, caller, shares);
+        }
+
+        _burn(owner, shares);
+
+        IERC20 asset = IERC20(asset());
+        if (fee > 0) {
+            asset.safeTransfer(registry.owner(), fee);
+        }
+        asset.safeTransfer(receiver, assets);
+
+        emit Withdraw(caller, receiver, owner, assets, shares);
+    }
+
+    function _beforeTokenTransfer(
+        address from,
+        address to,
+        uint256 amount
+    ) internal virtual override {
+        if (to != address(0)) {
+            if (registry.isEmergencyPaused()) revert SystemInEmergencyPause();
+            // Cannot transfer or mint tokens to a user for which an active withdraw request exists
```

Comment:
i was wondering if we can allow that and keep track of the amount the user had when he did the withdraw request (and this way the withdraw would have a limit)

Reply:
I agree this would be better UX (esp bc ppl cannot receive shares if they have a withdrawrequest open, which is just shitty). But it would also add complexity and new edge cases to consider. So idk.

Reply:
well the case of transfer was not possible at all in v1, and i'm not sure how much it would be in use, if at all.. but doesn't seem that complex to keep amount to withdraw per user and add that to the maxWithdraw check, but sure there's more i don't think of. maybe not worth it